### ISSUE TITLE
No newline after expression that is at end of the line

### ISSUE THREAD

**Issue by zylom** [PebbleTemplates/pebble#125]
When an expression is at the end of a line, there seems to be no newline character to seperate that line with the next line.

``` javascript
<html>
  <body>
    <div id="content">
      {{ mainContent }}
    </div>
  </body>
</html>
```

In the previous example, given that the `mainContent` variable will be replaced with `"Hello, world"`, the output will be:

``` html
<html>
  <body>
    <div id="content">
      Hello, world    </div>
  </body>
</html>
```

As you can see, there is no newline after `"Hello, world"`.  The `</div>` closing tag with its four spaces of indentation should be on the next line, not on the same line that `"Hello, world"` is on.

**Comment by mbosecke** [comment]
This is a fairly common complaint but the original intention of removing the newline after every pebble expression is because otherwise you would have a lot of blank lines. For example:

    <p>
    {%  if true %}
        lorem ipsum
    {% endif %}
    </p>

would print with two blank lines:
   
    <p>
    
        lorem ipsum
    
    </p>

I'm not necessarily opposed to change but we will want to make sure we maintain the above requirement.

**Comment by r0b0tbuilder** [comment]
To comply with that requirement, I was thinking of implementing this enhancement as follows:

Newline characters are removed only if that newline is after an expression, otherwise (such as in the case when it is at the end of a control-flow delimiter) the newline will be preserved.

If that sounds good, I'll start working on a pull request. :+1:

**Comment by mbosecke** [comment]
Let me just reword what you're saying to make sure we're both on the same page:

If the newline immediately follows a {% ... %} statement then it will be removed but if it follows a {{ ... }} statement it will be preserved. Correct?

I would agree that this is a step in the right direction and I would probably accept a pull request for this, although be forewarned that it may not be an easy task (although it's been awhile since I looked at the relevant code). 

I do also want to mention that we can optionally take this even a step further. Ideally, the newline would be preserved in the following example too even though it follows a {% ... %} statement:

```
Subject: {% if true %} subject line {% endif %}
Message: message
```

I think the ideal rule would be to only remove newlines if that entire line of text contains nothing but a single {% ... %} statement (and whitespace).

**Comment by r0b0tbuilder** [comment]
Yes, that is what I was trying to convey. We're both on the same page :smiley: . I agree with your suggestion, and I will try my best to implement this as you suggested.

**Comment by r0b0tbuilder** [comment]
I just realized that many of the unit tests will fail as many newlines will be kept, when it is expected for those newlines to be removed. Would you like me to edit the unit tests?

**Comment by mbosecke** [comment]
Shoot, I didn't even think about that but yes the unit tests will need to be updated so that they all pass.

**Comment by mbosecke** [comment]
Hey, did you ever have any luck with this? Perhaps I can help you finish it?

**Comment by r0b0tbuilder** [comment]
Sorry for the holdup; I should have a pull request submitted by the end of this week at the latest.

Here is what I found that is causing the newlines to be removed: Regexes (initalized in the constructor of the Syntax class) used to find particular delimiting tags (e.g. %} or }}) are matching a possible newline at the end of the line, which causes the Lexer to advance past the newline when it tokenizes the delimiter token. So the newline doesn't end up in a text token like it should. Does that make sense?

**Comment by basilmusa** [comment]
Is there an option to disable tampering with new line characters at end of variable completely?
I'm using Pebble for generating template bash scripts on Linux. I need to preserve new lines.
Looking for a method that would disable new line trimming completely, something like 

    PebbleEngine#enableNewLineTrmming(boolean enable).

Example:

    PebbleEngine engine = new PebbleEngine.Builder().enableNewLineTrimming(false).build();

**Comment by agilej** [comment]
Agree with @basilmusa , with one enable or disable config, we can generate different content under different environments such as development or production.

**Comment by CKovar** [comment]
since {{- someVariable -}} is hungry to consume additional whitespace, would it be reasonable to use something like {{+ someVariable +}} to explicitly preserve whitespace?

example usage:
```
here is line one, ending in {{+ someVariable +}}
here is line two.
```

if someVariable = "hello", the output would be:
```
here is line one, ending in hello
here is line two.
```

Thanks,  (great library BTW)

**Comment by electrotype** [comment]
I created a Pull Request for that feature : https://github.com/PebbleTemplates/pebble/pull/226

I only implemented the easy solution though, which is the one suggested by @basilmusa  : a `.enableNewLineTrimming()` method available when building a `PebbleEngine` instance and that can be used to disable _all_ trimming of newlines after Pebble tags. 

Tests are provided but some more would be ideal if you think of cases not covered. Also, this is the first time I look at Pebble source, so I hope I'm not missing something obvious!

Feel free to clean part of my code or correct my (not perfect) english too.
